osu.Game/Online/SignalRUnionWorkaroundResolver.cs

```diff
@@ -20,34 +19,22 @@ public class SignalRUnionWorkaroundResolver : IFormatterResolver
         public static readonly MessagePackSerializerOptions OPTIONS =
             MessagePackSerializerOptions.Standard.WithResolver(new SignalRUnionWorkaroundResolver());
 
-        public static readonly IReadOnlyList<Type> BASE_TYPES = new[]
-        {
-            typeof(MatchServerEvent),
-            typeof(MatchUserRequest),
-            typeof(MatchRoomState),
-            typeof(MatchUserState),
-        };
-
-        public static readonly IReadOnlyList<Type> DERIVED_TYPES = new[]
-        {
-            typeof(ChangeTeamRequest),
-            typeof(TeamVersusRoomState),
-            typeof(TeamVersusUserState),
-        };
+        private static readonly IReadOnlyDictionary<Type, IMessagePackFormatter> formatter_map = createFormatterMap();
 
-        private static readonly IReadOnlyDictionary<Type, IMessagePackFormatter> formatter_map = new Dictionary<Type, IMessagePackFormatter>
+        private static IReadOnlyDictionary<Type, IMessagePackFormatter> createFormatterMap()
         {
-            { typeof(TeamVersusUserState), new TypeRedirectingFormatter<TeamVersusUserState, MatchUserState>() },
-            { typeof(TeamVersusRoomState), new TypeRedirectingFormatter<TeamVersusRoomState, MatchRoomState>() },
-            { typeof(ChangeTeamRequest), new TypeRedirectingFormatter<ChangeTeamRequest, MatchUserRequest>() },
+            IEnumerable<(Type derivedType, Type baseType)> baseDerived = SignalRWorkaroundTypes.BASE_DERIVED;
 
-            // These should not be required. The fallback should work. But something is weird with the way caching is done.
+            // This should not be required. The fallback should work. But something is weird with the way caching is done.
             // For future adventurers, I would not advise looking into this further. It's likely not worth the effort.
-            { typeof(MatchUserState), new TypeRedirectingFormatter<MatchUserState, MatchUserState>() },
-            { typeof(MatchRoomState), new TypeRedirectingFormatter<MatchRoomState, MatchRoomState>() },
-            { typeof(MatchUserRequest), new TypeRedirectingFormatter<MatchUserRequest, MatchUserRequest>() },
-            { typeof(MatchServerEvent), new TypeRedirectingFormatter<MatchServerEvent, MatchServerEvent>() },
-        };
+            baseDerived = baseDerived.Concat(baseDerived.Select(t => (t.baseType, t.baseType))).Distinct();
```

Comment:
This reads weird. Shouldn't it be something like `baseDerived.Keys.Distinct().Select(t => (t, t))`?

Reply:
`baseDerived` still holds the entire list of tuples (or key/value pairs if I changed it to that) so not sure how that would work. Have renamed it to `baseMap` inline with the other suggestion in case that clears it up.

Reply:
You've removed the `Distinct` altogether now, which I guess is okay if it supports adding the same type more than once? I don't know.

Reply:
That distinct was for the `(typeof(MatchServerEvent), typeof(MatchServerEvent))` case anyways, which shouldn't really be a thing so I removed it now.

Reply:
Keep in mind that going forward, there will likely be more than one derived class. That's the whole point of making these in the first place. So it will happen.